File: packages/stow-py/src/stow_s3/bin.py
"""Locating the native stow server binary.

The resolution order and the shape of the result mirror the TypeScript client so
a user moving between the two languages sees the same behaviour and the same
diagnostic. The source matters as much as the path: "nothing is installed" and
"STOW_BIN points at a binary from an older release" are different problems with
different fixes, and a bare path cannot tell them apart.
"""

from __future__ import annotations

import os
import shutil
from dataclasses import dataclass
from enum import Enum
from pathlib import Path

#: Name of the binary bundled in a platform wheel, relative to the package.
BUNDLED_SUBPATH = Path("_bin") / "stow"

#: How far above the package to look for a monorepo checkout.
MONOREPO_DEPTH = 8


class BinarySource(str, Enum):
    """Which rule produced the resolved path."""

    PLATFORM_WHEEL = "platform-wheel"
    ENVIRONMENT = "environment"
    MONOREPO = "monorepo"
    PATH = "path"


@dataclass(frozen=True)
class ResolvedBinary:
    path: str
    source: BinarySource


class StowBinaryNotFoundError(Exception):
    """No runnable server binary was found.

    The underlying spawn failure for a missing binary is a bare ``ENOENT`` on the
    literal string ``stow``, which tells someone who installed from PyPI nothing
    about why their session did not start.
    """

    code = "binary_not_found"

    def __init__(self, resolved: str) -> None:
        super().__init__(
            f"The stow server binary was not found (resolved to {resolved!r}). "
            "It is searched for in this order: the platform wheel that ships it "
            "inside this package, the STOW_BIN environment variable, bin/stow "
            "relative to a monorepo checkout, then stow on PATH. Install stow-s3 "
            "for this platform, point STOW_BIN at an existing binary, or set "
            "STOW_BIN to a downloaded release."
        )
        self.resolved = resolved
        self.name = "StowBinaryNotFoundError"


def _is_executable(path: Path) -> bool:
    # A path that exists but is not executable is worse than one that is absent,
    # because a directory can satisfy exists() and only fail at spawn.
    return path.is_file() and os.access(path, os.X_OK)


def _bundled_binary() -> Path | None:
    """The binary shipped inside a platform wheel, if this is one."""
    candidate = Path(__file__).resolve().parent / BUNDLED_SUBPATH
    return candidate if _is_executable(candidate) else None


def _monorepo_binary() -> Path | None:
    directory = Path(__file__).resolve().parent
    for _ in range(MONOREPO_DEPTH):
        if (directory / "go.mod").is_file():
            candidate = directory / "bin" / "stow"
            if _is_executable(candidate):
                return candidate
        parent = directory.parent
        if parent == directory:
            break
        directory = parent
    return None


def resolve_stow_binary_detailed() -> ResolvedBinary:
    """Resolve the binary path and report which rule produced it."""
    bundled = _bundled_binary()
    if bundled is not None:
        return ResolvedBinary(str(bundled), BinarySource.PLATFORM_WHEEL)

    from_env = (os.environ.get("STOW_BIN") or "").strip()
    if from_env:
        return ResolvedBinary(from_env, BinarySource.ENVIRONMENT)

    in_repo = _monorepo_binary()
    if in_repo is not None:
        return ResolvedBinary(str(in_repo), BinarySource.MONOREPO)

    return ResolvedBinary("stow", BinarySource.PATH)


def resolve_stow_binary() -> str:
    return resolve_stow_binary_detailed().path


def stow_binary_available() -> bool:
    """Whether a resolved path is actually runnable."""
    resolved = resolve_stow_binary_detailed()
    if resolved.source is BinarySource.PATH:
        return shutil.which("stow") is not None
    return _is_executable(Path(resolved.path))


def require_stow_binary() -> ResolvedBinary:
    """Resolve the binary, raising an actionable error when it is not runnable."""
    resolved = resolve_stow_binary_detailed()
    if not stow_binary_available():
        raise StowBinaryNotFoundError(resolved.path)
    return resolved


def describe_source(source: BinarySource) -> str:
    """A human explanation of a resolution source."""
    return {
        BinarySource.PLATFORM_WHEEL: "the platform wheel that ships it in this package",
        BinarySource.ENVIRONMENT: "the STOW_BIN environment variable",
        BinarySource.MONOREPO: "bin/stow in the monorepo checkout",
        BinarySource.PATH: "stow on PATH",
    }[source]

File: packages/stow/src/bin.ts
import { accessSync, constants, existsSync } from "node:fs";
import { delimiter, dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const PACKAGE_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");

function isExecutable(path: string): boolean {
  try {
    accessSync(path, constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

function findOnPath(name: string): string | undefined {
  const extensions =
    process.platform === "win32"
      ? (process.env.PATHEXT ?? ".EXE;.CMD;.BAT").split(";")
      : [""];
  for (const directory of (process.env.PATH ?? "").split(delimiter)) {
    if (!directory) {
      continue;
    }
    for (const extension of extensions) {
      const candidate = join(directory, `${name}${extension.toLowerCase()}`);
      if (isExecutable(candidate)) {
        return candidate;
      }
      if (process.platform !== "win32") {
        break;
      }
    }
  }
  return undefined;
}

function findMonorepoBinary(): string | undefined {
  let dir = PACKAGE_ROOT;
  for (let depth = 0; depth < 8; depth += 1) {
    if (existsSync(join(dir, "go.mod"))) {
      const candidate = join(dir, "bin", "stow");
      if (isExecutable(candidate)) {
        return candidate;
      }
    }
    const parent = dirname(dir);
    if (parent === dir) {
      break;
    }
    dir = parent;
  }
  return undefined;
}

/**
 * Resolve the stow binary path.
 *
 * Precedence: STOW_BIN env, `bin/stow` relative to monorepo root, then `stow` on PATH.
 */
export function resolveStowBinary(): string {
  const fromEnv = process.env.STOW_BIN?.trim();
  if (fromEnv) {
    return fromEnv;
  }

  const fromRepo = findMonorepoBinary();
  if (fromRepo) {
    return fromRepo;
  }

  return "stow";
}

export function stowBinaryAvailable(): boolean {
  const bin = resolveStowBinary();
  if (bin !== "stow") {
    return isExecutable(bin);
  }
  return findOnPath("stow") !== undefined;
}

/**
 * StowBinaryNotFoundError reports that no runnable server binary was found.
 * It exists because the underlying spawn failure is a bare ENOENT for the
 * literal string "stow", which tells someone who installed from npm nothing
 * about why their session did not start or what to do about it.
 */
export class StowBinaryNotFoundError extends Error {
  readonly code = "binary_not_found";

  constructor(resolved: string) {
    super(
      `The stow server binary was not found (resolved to ${JSON.stringify(resolved)}). ` +
        "It is searched for in this order: the STOW_BIN environment variable, " +
        "bin/stow relative to a monorepo checkout, then stow on PATH. " +
        "Install the platform binary package for this platform, point STOW_BIN at an " +
        "existing binary, or use the EmbeddedStow and @chs/stow/browser profiles, " +
        "which need no server binary.",
    );
    this.name = "StowBinaryNotFoundError";
  }
}
